### PR TITLE
Stackingcvclassifier

### PR DESCRIPTION
**Pull request by reiinakano**
Please make sure that these boxes are checked before submitting a new pull request -- thank you!

- [x] Create and check out a new topic branch
- [x] Implement the new feature or apply the bug-fix
- [x] Add appropriate unit test functions in the `./mlxtend/*/tests` directories
- [x] Run `nosetests ./mlxtend -sv` and make sure that all unit tests pass
- [x] Check/improve the test coverage by running `nosetests ./mlxtend --with-coverage`
- [x] Check for style issues by running `flake8 ./mlxtend` (you may want to run nosetests again after you made modifications to the code)
- [x] Add a note about the modification or contribution to the `./docs/sources/`CHANGELOG.md` file
- [ ] Modify documentation in the appropriate location under `mlxtend/docs/sources/`
- [x] Push the topic branch to the server and create a pull request
- [x] Check the Travis-CI build passed at https://travis-ci.org/rasbt/mlxtend
- [x] Squash (many small) commits to a larger commit

For more information and instructions, please see http://rasbt.github.io/mlxtend/contributing/

### CODE REVIEW

**Comment by rasbt** [comment]
Thanks for the contribution so far! Just see that there's an issue with travis and matplotlib coming up now. Hopefully, that's fixed soon -- it worked until 1 week ago, I guess they made some changes to their "trusty" distribution, which was still beta. Will try to switch back to the travis default distro since tensorflow should be available via conda as of today.

**Comment by rasbt** [comment]
Just moved the plotting functions over to `mlxtend.plotting`. Something I wanted to do anyway at some point (#84). Still don't understand why both Appveyor and TravisCI broke at the same time with regard to matplotlib, since it worked fine just one week ago: #98. Probably due to some changes in conda or so. In any case, tried various things, but wasn't really able to figure it out. The matplotlib functions inside the other subpackages were annoying anyways (as users also pointed out).

I will squash & merge this pull request (#75) into the master branch tomorrow so that you can rebase and try the tests again.

**Comment by reiinakano** [comment]
Okay thanks. :)

I have a question though. I'm new to this so I'm not sure but what exactly are the tests for and how do I make sure I don't break them?

**Comment by rasbt** [comment]
> I have a question though. I'm new to this so I'm not sure but what exactly are the Appveyor and TravisCI tests for and how do I make sure I don't break them?

No worries, these are just automated test and you can think of them as running `nosetests` locally. It's a check for me that everything runs correctly, but it also runs the unittests on different systems (e.g., AppVeyor runs on Windows and Travis CI runs on Linux). Also, the Travis checks make sure that the code works for both Python 2.7 and 3.5. If you are interested, the configuration file is here: https://github.com/rasbt/mlxtend/blob/master/.travis.yml

I just merged the pull request that moved plotting functions to `mlxtend.plotting`. I guess you need to do a rebase now, which basically means that you apply the changes from the master branch to your branch, and put your changes on top of it. 

- I.e., you would have to sync your repo with the upstream master branch first: see info for syncing here: https://rasbt.github.io/mlxtend/contributing/#syncing-an-existing-fork
- then `git checkout <your branch>`
- then  `git rebase upstream/master`

Thanks again for your contribution, I will take a look at the actual code soon!

**Comment by reiinakano** [comment]
Sorry for all the weird pulls. I still don't understand very well how Git works.

I followed the steps you mentioned about doing a rebase, and successfully did the rebase on my local repository. After that, I tried pushing the changes into my remote Github repository but it didn't work. I guess there are other steps I missed to do this properly but I couldn't figure them out. After a series of random pulls involving my remote and local repositories, I somehow managed to do a successful push from my local repo. At this point, though, I'm probably better off deleting my local repository and starting from scratch. I'm sure I've badly screwed up my Git tree somewhere along the way.

Also, sorry for the small commits afterwards. Was fixing some bugs so it would work in Python 3. I know I'm supposed to squash them into a smaller commit, but I couldn't figure out how to do that after I've already pushed the commits into my remote repository.

**Comment by rasbt** [comment]
No worries regarding the many small commits! Also, no need to worry about the squashing anymore, since GitHub added a new, handy feature a few weeks ago that allows me to do the squash conveniently at the very end (I updated the pull request template accordingly):

screen shot 2016-09-20 at 1 49 39 pm
Will try to provide some feedback later when I get home or tomorrow! Thanks!

File: mlxtend/classifier/stacking_cv_classification.py [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+        self : object
+
+        """
+        self.clfs_ = [clone(clf) for clf in self.classifiers]
+        self.meta_clf_ = clone(self.meta_classifier)
+        if self.verbose > 0:
+            print("Fitting %d classifiers..." % (len(self.classifiers)))
+
+        if self.stratify:
+            skf = StratifiedKFold(y, n_folds=self.n_folds)
```
Comment (rasbt) [inline mlxtend/classifier/stacking_cv_classification.py]: I'd say the random_state for both KFold and StratifiedKFold would need  to be set here. We'd need to add another __init__ parameter, random_state then, defaulting to `random_state=None`.

File: mlxtend/classifier/stacking_cv_classification.py [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+                if self.verbose > 0:
+                    print ("Training and fitting fold %d of %d..." % \
+                          ((num+1), self.n_folds))
+
+                if not self.use_probas:
+                    prediction = model.fit(X[train_index], y[train_index])\
+                        .predict(X[test_index])
+                    prediction = prediction.reshape(prediction.shape[0], 1)
+                else:
+                    prediction = model.fit(X[train_index], y[train_index]).\
```
Comment (rasbt) [inline mlxtend/classifier/stacking_cv_classification.py]: This could be slightly refactored by fitting the model one time before the if else clause and then just call `model.predict` or `model.predict_proba` respectively

File: mlxtend/classifier/stacking_cv_classification.py [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+from sklearn.cross_validation import KFold
+
+
+class StackingCVClassifier(BaseEstimator, ClassifierMixin, TransformerMixin):
+
+    """A Stacking CV classifier for scikit-learn estimators for classification.
+
+    Parameters
+    ----------
+    classifiers : array-like, shape = [n_regressors]
```
Comment (rasbt) [inline mlxtend/classifier/stacking_cv_classification.py]: I think this should say `shape = [n_classifiers]` not `shape = [n_regressors]` :)

File: docs/sources/CHANGELOG.md [hunk 4]
```diff
@@ -14,6 +14,7 @@ The CHANGELOG for the current development version is available via
 ##### New Features
 
 - New `ExhaustiveFeatureSelector` estimator in `mlxtend.feature_selection` for evaluating all feature combinations in a specified range
+- New 'StackingCVClassifier' estimator in 'mlxtend.classifier' for implementing a stacking ensemble that uses cross-validation techniques for training the meta-estimator to avoid overfitting.
```
Comment (rasbt) [inline docs/sources/CHANGELOG.md]: I'd suggest you add your name or GitHub handle to this line. E.g., [Reiichiro Nakano](https://github.com/reiinakano)

File: mlxtend/classifier/stacking_cv_classification.py [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+    Attributes
+    ----------
+    clfs_ : list, shape=[n_classifiers]
+        Fitted classifiers (clones of the original classifiers)
+    meta_clf_ : estimator
+        Fitted meta-classifier (clone of the original meta-estimator)
+
+    """
+    def __init__(self, classifiers, meta_classifier,
+                 use_probas=True, n_folds=2,
```
Comment (rasbt) [inline mlxtend/classifier/stacking_cv_classification.py]: `use_probas` should be set to `use_probas=False` in this case; otherwise, it would be inconsistent with the docstring. I think `use_probas=False` may be better than `use_probas=True` since some scikit estimators may not support predict_proba yet.

**Comment by rasbt** [comment]
I added a few inline comments above. Would be nice if you could address them some time before I look more into the code so that it doesn't become too confusing. So far, it looks good overall though!

**Comment by reiinakano** [comment]
Alright. Thanks for the comments. I'll start working on them now. :)

File: mlxtend/classifier/stacking_cv_classification.py (hunk 1 above)
Comment (reiinakano) [inline mlxtend/classifier/stacking_cv_classification.py]: I have a question about this. The `random_state` parameter is only used when the `shuffle` parameter is True. Does that mean that if  `random_state` is None, we go ahead and make `shuffle` False?

File: docs/sources/CHANGELOG.md [hunk 6]
```diff
@@ -15,7 +15,7 @@ The CHANGELOG for the current development version is available via
 
 - New `ExhaustiveFeatureSelector` estimator in `mlxtend.feature_selection` for evaluating all feature combinations in a specified range
 - The `StackingClassifier` has a new parameter `average_probas` that is set to `True` by default to maintain the current behavior. A deprecation warning was added though, and it will default to `False` in future releases (0.5.0); `average_probas=False` will result in stacking of the level-1 predicted probabilities rather than averaging these.
-- New 'StackingCVClassifier' estimator in 'mlxtend.classifier' for implementing a stacking ensemble that uses cross-validation techniques for training the meta-estimator to avoid overfitting.
+- New 'StackingCVClassifier' estimator in 'mlxtend.classifier' for implementing a stacking ensemble that uses cross-validation techniques for training the meta-estimator to avoid overfitting. <github.com/reiinakano>
```
Comment (rasbt) [inline docs/sources/CHANGELOG.md]: Sorry about that again, but the changelog gets converted to Markdown for the online documentation at http://rasbt.github.io/mlxtend/changelog/, so I'd suggest `[Reiichiro Nakano](https://github.com/reiinakano)` or `[@reiinakano] (https://github.com/reiinakano)`. This way it would appear as a link so that people can click to it to get to your account if you prefer that.

File: mlxtend/classifier/stacking_cv_classification.py [hunk 7]
```diff
@@ -138,16 +139,16 @@ def fit(self, X, y):
             for num, (train_index, test_index) in enumerate(skf):
 
                 if self.verbose > 0:
-                    print ("Training and fitting fold %d of %d..." % \
-                          ((num+1), self.n_folds))
+                    print ("Training and fitting fold %d of %d..." %
+                           ((num+1), self.n_folds))
+
+                prediction = model.fit(X[train_index], y[train_index])
```
Comment (rasbt) [inline mlxtend/classifier/stacking_cv_classification.py]: Here, you could just call `model.fit(X[train_index], y[train_index])` without assigning it to a parameter and use `prediction = model.predict(X[test_index])`, which would be a bit clearer in debugging

File: mlxtend/classifier/stacking_cv_classification.py (hunk 1 above)
Comment (rasbt) [inline mlxtend/classifier/stacking_cv_classification.py]: Good point! I think we should definitely set `shuffle=True` by default (and add a shuffle `__init__` parameter). Otherwise this may cause weird results if people didn't shuffle their datasets.

**Comment by rasbt** [comment]
Wanted to let you know that I deleted some of the older comments that you already addressed to keep track of what's already been completed. GitHub's UI has changed recently introducing this "review" feature and I have to play around with it more to figure out a better workflow for this :P

**Comment by rasbt** [comment]
This looks very good to me and wouldn't have any suggestions at this point. I'd be happy to merge if you don't plan on any further changes. I can take care of the documentation then since this is maybe a bit tricky at the current moment -- I wanted to make some explanatory figures on StackingCV anyways. I can ping you then for feedback once I have the pull request ready.

**Comment by reiinakano** [comment]
That's good to hear. I don't think I have any further changes in mind right now. :)
